README.md


# Static site  and  styleguide Maker
Description
===========
Static site  and  styleguide Maker

Requirements
------------
* Node
* NPM
* Grunt
* Ruby
* SASS 3.4+
*Check SASS version sass-v
*Assemble http://assemble.io/

Folder Structure
----------------
```
/src            // Main Folder.
--/assets         // assets
----/css         //where your sass files get parsed to .
-----/styleguide.md         //This file MUST be here for the styleguide to work .
----/images         // .
-----/sprites         // put all your sprites here. Grunt will build spritesheet.png for you. It will also add a _sprites.scss file in your scss folder
----/js         // .
--/content         // .
----/markdown.md         // .
--/data         // .
--/scss         //put all your sass files here . 
--/styleguide         // the template for the styleguide. You can change the look of the styleguide here .
--/templates         //your hbs files for assemble go here .
----/layout         // .
----/pages         // .
----/partials         // .
/dist            // folder created when you run grunt server
/node_modules     // Folder created when running  npm install
/bower_components     // Folder created when running  bower install


```


Install node.js
---------------
Reference: http://nodejs.org/
Go to the node.js website and click the Install button.

Check to see if node.js installed properly...

```

$ node -v
```


Install grunt.js
----------------
Reference: http://gruntjs.com/getting-started#installing-the-cli

```

$ sudo npm install -g grunt-cli
```


Install bower
----------------
Reference: http://bower.io/#install-bower

```

$ npm install -g bower



Install node.js Dependencies / Modules
--------------------------------------
From within the cloned repository...

```
$ npm install
```


Install SASS
----------------------------
Reference: http://sass-lang.com/install
From within the cloned repository...

```
$ gem install sass
```

Using browsersync
------------------------
Reference: hhttp://www.browsersync.io/

Using Assemble
------------------------
http://assemble.io/docs/

start the server to preview static pages
------------------------
install bower components: bower install
install nmp dependencies: sudo npm install

Preview an app you have generated (with browsersync).
$ grunt server
or with more info on the terminal.
$ grunt server --verbose 

Using Jshint
------------------------
```
if you add any plugins for jquery add this to the top of each file
so jshint will ignore them
// jshint ignore: start

Also if you you want you can edit the jshintrc file and add the plugin vars to the 'globals' array. 
Example
  "globals": {
    "$":false,
    "jquery":false,
    "app":false,
    "google":false,
    "Modernizr":false
        // other explicit global names to exclude
  }
```

Dependencies already in the layout/defualt.hbs
------------------------
```
modernizr.js v2.8.3
You can upgrade to 3 if  desired. Just do
bower uninstall  modernizr

and reinstall 3 

or remove this dependency "bower uninstall  modernizr" and add it manually.

 jQuery v1.11.3
 Is also included by default. Do the same as in modernizr if you want a different version of jquery

```

Importing CSS Breakpoints Into JavaScript
------------------------
```
https://www.lullabot.com/articles/importing-css-breakpoints-into-javascript

I have setup   _mediaquery-reference.scss and in apps.js so you can check in your JS what breakpoint you are in. I have setup a simple example


body:after {
	content: "mobile";
	display: none;
}
@media only screen and (min-width: 1025px) {
	body:after {
		content: "desktop";
		
	}
}
@media only screen and (min-width: 1200px) {
	body:after {
		content: "bigdesktop";
		
	}
}


```



CHANGELOG
v0.1.0:
  date: "2015-09-15"
  changes:
    - Generated by the [Yeoman Generator](https://github.com/assemble/generator-assemble) for [Assemble](http://assemble.io)


src/styleguide/public/kss.js
(function() {
  var KssStateGenerator;

  KssStateGenerator = (function() {

    function KssStateGenerator() {
      var idx, idxs, pseudos, rulePattern, ruleMatches, replaceRule, rule, stylesheet, _i, _r, _len, _len2, _len3, _ref, _ref2;
      rulePattern = /[^|,]([a-z0-9\.\- >]+(\:hover|\:disabled|\:active|\:visited|\:focus)[a-z0-9\.\- >]*)/gi;
      pseudos = /(\:hover|\:disabled|\:active|\:visited|\:focus)/g;
      _ref = document.styleSheets;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        stylesheet = _ref[_i];
        idxs = [];
        _ref2 = stylesheet.cssRules || [];
        for (idx = 0, _len2 = _ref2.length; idx < _len2; idx++) {
          rule = _ref2[idx];
          if ((rule.type === CSSRule.STYLE_RULE) && pseudos.test(rule.selectorText)) {
            replaceRule = function(matched, stuff) {
              return ".pseudo-class-" + matched.replace(':', '');
            };
            ruleMatches = rule.cssText.match(declaration);
            for (_r = 0, _len3 = ruleMatches.length; _r < _len3; _r++) {
              this.insertRule(ruleMatches[_r].replace(pseudos, replaceRule));
            }
          }
        }
      }
    }

    KssStateGenerator.prototype.insertRule = function(rule) {
      var headEl, styleEl;
      headEl = document.getElementsByTagName('head')[0];
      styleEl = document.createElement('style');
      styleEl.type = 'text/css';
      if (styleEl.styleSheet) {
        styleEl.styleSheet.cssText = rule;
      } else {
        styleEl.appendChild(document.createTextNode(rule));
      }
      return headEl.appendChild(styleEl);
    };

    return KssStateGenerator;

  })();

  new KssStateGenerator;

}).call(this);




Gruntfile.js
/*
 * Generated on 2015-09-15
 * generator-assemble v0.5.0
 * https://github.com/assemble/generator-assemble
 *
 * Copyright (c) 2015 Hariadi Hinta
 * Licensed under the MIT license.
 */

'use strict';

// # Globbing
// for performance reasons we're only matching one level down:
// '<%= config.src %>/templates/pages/{,*/}*.hbs'
// use this if you want to match all subfolders:
// '<%= config.src %>/templates/pages/**/*.hbs'

module.exports = function(grunt) {

	require('time-grunt')(grunt);
	require('load-grunt-tasks')(grunt);

	// Project configuration.
	grunt.initConfig({

		config: {
			src: 'src',
			dist: 'dist'
		},

		watch: {
			bower: {
				files: ['bower.json'],
				tasks: ['wiredep']
			},
			assemble: {
				files: ['<%= config.src %>/{content,data,templates}/{,*/}*.{md,hbs,yml}'],
				tasks: ['assemble','clean:style','styleguide']
			},
			sass: {
				files: ['<%= config.src %>/scss/{,*/}*.{scss,sass}'],
				tasks: ['sass:server', 'autoprefixer','clean:style','styleguide']
			},
			css: {
			files: ['<%= config.src %>/scss/{,*/}*.{css}'],
			tasks: ['assemble','kss']
			},
			js: {
				files: ['<%= config.src %>/assets/js/{,*/}*.js'],
				tasks: ['jshint','copy:js']
				
			},

			sprite:{
				files: ['<%= config.src %>/assets/images/sprites/*.png'],
				tasks: ['sprite']
			}

		},
		 styleguide: {
			build: {
				options: {
					
					template: {
						src: '<%= config.src %>/styleguide/'
						//kss template contained in this directory so that it can be editable
					},
					framework: {
						name: 'kss'
						
					}
				},
				files: {
					 '<%= config.dist %>/styleguide': ['<%= config.src %>/assets/css/main.css']
				}
			}
		},
		assemble: {
			
			pages: {
				options: {
					flatten: true,
					assets: '<%= config.dist %>/assets',
					layout: '<%= config.src %>/templates/layouts/default.hbs',
					data: '<%= config.src %>/data/*.{json,yml}',
					partials: '<%= config.src %>/templates/partials/*.hbs',
					plugins: ['assemble-contrib-permalinks','assemble-contrib-sitemap'],
				},
				files: {
					'<%= config.dist %>/': ['<%= config.src %>/templates/pages/*.hbs']
				}
			},
			options: {
					plugins: ['assemble-middleware-kssnode']
				}
		},

		copy: {

			dist: {
				expand: true,
				cwd: '<%= config.src %>/assets/',
				src: ['**/*','!*.txt','!*.md','!**/sprites/**'],
				dest: '<%= config.dist %>/assets/'
			},
			js: {
				expand: true,
				cwd: '<%= config.src %>/assets/js/',
				src: ['**/*','!*.txt','!*.md','!**/sprites/**'],
				dest: '<%= config.dist %>/assets/js/'
			}
			
		},
		modernizr: {
			dist: {
				devFile: 'bower_components/modernizr/modernizr.js',
				outputFile: '<%= config.dist %>/assets/js/vendor/modernizr.js',
				files: {
					src: [
						'<%= config.dist %>/scripts/{,*/}*.js',
						'<%= config.dist %>/styles/{,*/}*.css',
						'!<%= config.dist %>/scripts/vendor/*'
					]
				},
				uglify: true
			}
		},
		// wiredep: {
		// 	dist: {
		// 		ignorePath: /^\/|\.\.\//,
		// 		src: ['<%= config.dist %>/*.html']
		// 	}
			
		// },
		// Compiles Sass to CSS and generates necessary files if requested
		sass: {
			dist: {
				files: [{
					expand: true,
					cwd: '<%= config.src %>/assets/scss',
					src: ['*.{scss,sass}'],
					dest: '<%= config.dist %>/assets/css/',
					ext: '.css'
				}]
			},
			server: {
				files: [{
					expand: true,
					cwd: '<%= config.src %>/scss',
					src: ['*.{scss,sass}'],
					dest: '<%= config.src %>/assets/css/',
					ext: '.css'
				}]
			}
		},
				// Reads HTML for usemin blocks to enable smart builds that automatically
		// concat, minify and revision files. Creates configurations in memory so
		// additional tasks can operate on them
		useminPrepare: {
			options: {
				dest: '<%= config.dist %>',
				flow: {
					steps: {
						js: ['concat'],
						css: ['concat']
					},
					post: {}
				}
			},
			html: '<%= config.dist %>/{,*/}*.html',
		},
		// Performs rewrites based on rev and the useminPrepare configuration
		usemin: {
			options: {
				assetsDirs: [
					'<%= config.dist %>',
					'<%= config.dist %>/images',
					'<%= config.dist %>/assets/css/',
					'<%= config.dist %>/assets/js/'
				]
			},
			html: ['<%= config.dist %>/{,*/}*.html'],
			css: ['<%= config.dist %>/assets/css/{,*/}*.css'],
			js: ['<%= config.dist %>/assets/js/{,*/}*.js']
		},
		uglify: {
			dist: {
				files: {
					'<%= config.dist %>/assets/js/{,*/}*.js': ['<%= config.dist %>/assets/js/{,*/}*.js']
				}
			}
			
		},
		
		concat: {
			dist: {}
		},
		htmlmin: {
			dist: {
				options: {
				 /* collapseBooleanAttributes: true,
					collapseWhitespace: true,
					conservativeCollapse: true,
					removeAttributeQuotes: true,
					removeCommentsFromCDATA: true,
					removeEmptyAttributes: true,
					removeOptionalTags: true,
					removeRedundantAttributes: true,
					useShortDoctype: true*/
				},
				files: [{
					expand: true,
					cwd: '<%= config.dist %>',
					src: '{,*/}*.html',
					dest: '<%= config.dist %>'
				}]
			}
		},


		// Add vendor prefixed styles
		autoprefixer: {
			options: {
				browsers: ['> 1%', 'last 4 versions', 'Firefox ESR', 'Opera 12.1'],
				map: true
			},
			dist: {
				files: [{
					expand: true,
					cwd: '<%= config.dist %>/assets/css/',
					src: '{,*/}*.css',
					dest: '<%= config.dist %>/assets/css/'
				}]
			}
		},
		 sprite:{
			all: {
				src: '<%= config.src %>/assets/images/sprites/*.png',
				dest: '<%= config.src %>/assets/images/spritesheet.png',
				destCss: '<%= config.src %>/scss/_sprites.scss',
				imgPath: '../assets/images/spritesheet.png'
			}
		},
		// Make sure code styles are up to par and there are no obvious mistakes
		jshint: {
			options: {
				jshintrc: '.jshintrc',
				reporter: require('jshint-stylish')
			},
			all: [
				'Gruntfile.js',
				'<%= config.src %>/assets/js/{,*/}*.js'
				
			]
		},
		browserSync: {
			files: {
				src : [
					'<%= config.dist %>/{,*/}*.html',
					'<%= config.dist %>/assets/js/{,*/}*',
					'<%= config.dist %>/assets/css/',
					'<%= config.dist %>/assets/images/{,*/}*'
				]
			},
			options: {
				watchTask: true,
				reloadDelay: 2000,
				ghostMode: {
					scroll: true,
					links: true,
					forms: true,
					ui: {
						port: 3005
					}
				},
				server: {
					baseDir: [
					'<%= config.dist %>'
					],
					routes: {
						'/bower_components': './bower_components'
					}
				}
			}
		},
		// Before generating any new files,
		// remove any previously-created files.
		clean:{
			dist: {
				files: [{
					dot: true,
					src: [
						'<%= config.dist %>/*'
					]
				}]
			},
			style: '<%= config.dist %>/styleguide*'

		}


	});

	grunt.loadNpmTasks('assemble');

	grunt.registerTask('server', [
		'build',
		'browserSync',
		'watch'
	]);

	grunt.registerTask('build', [
		'clean:dist',
		'sass:server',
		'sprite',
		'copy',
		'assemble',
		'useminPrepare',
		'modernizr',
		'autoprefixer',
		'concat',
		'usemin',
		'htmlmin',
		//'uglify', Uncomment if you want to uglify your files
		'styleguide:build'
		
	]);

	grunt.registerTask('default', [
		'build'
	]);

};
